### DISCUSSION TITLE
Patching ACPI tables (SSDT)

### DISCUSSION THREAD

**Discussion by fermino** [zbm-dev/zfsbootmenu#500] [General]
Hi! First and foremost, thanks for all the effort put in this project. I have found it extremely useful!

I had a question regarding loading custom acpi tables. My notebook requires some patches for the sound to work and so far I have been loading them with grub: 
```
### BEGIN /etc/grub.d/01_acpi ###
acpi ($root)/ssdt-csc3551.aml
### END /etc/grub.d/01_acpi ###
```
Now that I have ditched grub, I'm trying to get the same thing working, so far I have found the following options, but as I do not have that much experience in the topic I'm getting a bit lost. 

- Patch the kernel (this is what I'm trying to avoid as I have more than one distro in the same zroot pool). 
- [Use a CPIO archive](https://wiki.archlinux.org/title/DSDT#Using_a_CPIO_archive). I understand I can concat the initrd file and my patch in one file, but here's when I get a bit lost as I don't completely understand how the boot process works. 
- Bundle it in a custom ZBM build: I do like this option as I would be avoiding patching all the distros, but I still would like to understand the previous method. 
- Patch ZBM so that it can load acpi tables on demand? It would be sort of mimicking grub's behavior; but I'm not sure if that can be done on the fly before kexec'ing the new kernel. 

What would be the most correct way to achieve this? I'm willing to put in the time for the last one if it is worth it, but I certainly need some hints :)

Thanks and kind regards!
Fermín Olaiz.

**Comment by ahesford** [comment]
I think the right approach is the second one: include your patched DSDT in a CPIO archive that is prepended to your actual initramfs. ZFSBootMenu will be completely unaware o fthe fact that your initramfs is actually two concatenated images, and load it the same way `systemd-boot` loads two images as suggested in the Arch wiki.

If you follow the Arch wiki steps to create the `acpi_override` image, it's as simple as

```sh
cp /boot/initramfs-[...].img /boot/initramfs-[...].img.bak
cat /path/to/acpi_override /boot/initramfs-[...].img.bak > /boot/initramfs-[...].img
```

This will leave the modified image where ZBM expects to find it, but the original initramfs alongside it as a backup if needed. You'll have to re-run this every time you create a new initramfs, *e.g.*, after a kernel upgrade.

Attempting to incorporate this capability into ZFSBootMenu may be challenging at best and, at worst, might not even work; I'm not sure if you can load replacement ACPI tables with ZFSBootMenu and have them persist after `kexec`.

**Reply by zdykstra** [reply]
A pre-kexec hook script could be used to automagically do the concat operation for the selected initramfs, and then subsequently load that image and the selected kernel into memory.

**Reply by fermino** [reply]
Thanks for the prompt response! After trying it seems that the acpi override only works for non-compressed initrd's, so I ended up using a custom mkinitcpio hook (arch) to prepend it to the compressed cpio and automate it for future updates: 

```bash
cat /etc/initcpio/post/acpi_override 
#!/usr/bin/env bash

ACPI_OVERRIDE_DIR="/etc/initcpio/acpi_override.d"

# As we can't override acpi tables using a compressed initcpio, we will prepend it to the image
if stat -t ${ACPI_OVERRIDE_DIR}/*.cpio >/dev/null 2>&1; then
	echo "    -> Prepending cpio images in ${ACPI_OVERRIDE_DIR}/*.cpio"

	mv "$2" "$2.bak"
	cat ${ACPI_OVERRIDE_DIR}/*.cpio "$2.bak" > "$2"
	rm "$2.bak"
fi
```
